File: examples/virtual_methods/virtual_methods.cpp
#include <monobind/monobind.h>

#include <iostream>

int main()
{
    const char* path_to_mono = MONOBIND_MONO_ROOT;
    monobind::mono mono(path_to_mono);
    mono.init_jit("VirtualMethods");

    // build csharp library
    monobind::compiler compiler(mono.get_root_dir());
    compiler.build_library("VirtualMethods.dll", "VirtualMethods.cs");
    
    // load assembly and get C# method
    monobind::assembly assembly(mono.get_domain(), "VirtualMethods.dll");

    monobind::object base_instance = assembly.get_method("MonoBindExamples.Factory::CreateBase()").as_function<monobind::object()>()();
    monobind::object derived1_instance = assembly.get_method("MonoBindExamples.Factory::CreateDerived1()").as_function<monobind::object()>()();
    monobind::object derived2_instance = assembly.get_method("MonoBindExamples.Factory::CreateDerived2()").as_function<monobind::object()>()();
    
    std::cout << "instances:" << std::endl;
    std::cout << base_instance << std::endl;
    std::cout << derived1_instance << std::endl;
    std::cout << derived2_instance << std::endl;
    std::cout  << std::endl;

    std::cout << "method calls:" << std::endl;
    base_instance.get_method<void()>("::Foo()")();
    derived1_instance.get_method<void()>("::Foo()")();
    derived2_instance.get_method<void()>("::Foo()")();

    return 0;
}

File: examples/virtual_methods/VirtualMethods.cs
using System;
using System.Runtime.CompilerServices;

namespace MonoBindExamples
{
    public class Base
    {
        public virtual void Foo()
        {
            Console.WriteLine("Called Base::Foo");
        }
    }

    public static class Factory
    {
        public static Base CreateBase()
        {
            return new Base();
        }

        public static Base CreateDerived1()
        {
            return new Derived1();
        }

        public static Base CreateDerived2()
        {
            return new Derived2();
        }
    }

    public class Derived1 : Base
    {
        public override void Foo()
        {
            Console.WriteLine("Called Derived1::Foo");
        }
    }

    public class Derived2 : Base
    {

    }
}